source/utilities/server.ts

```diff
@@ -0,0 +1,126 @@
+// source/utilities/server.ts
+// Run the server with the given configuration.
+
+import http from 'node:http';
+import https from 'node:https';
+import { readFile } from 'node:fs/promises';
+import handler from 'serve-handler';
+import compression from 'compression';
+import isPortReachable from 'is-port-reachable';
+import { getNetworkAddress, registerCloseListener } from './http.js';
+import { promisify } from './promise.js';
+import type { IncomingMessage, ServerResponse } from 'node:http';
+import type { AddressInfo } from 'node:net';
+import type {
+  Configuration,
+  Options,
+  ParsedEndpoint,
+  Port,
+  ServerAddress,
+} from '../types.js';
+
+const compress = promisify(compression());
+
+/**
+ * Starts the server and makes it listen on the given endpoint.
+ *
+ * @param endpoint - The endpoint to listen on.
+ * @param config - The configuration for the `serve-handler` middleware.
+ * @param args - The arguments passed to the CLI.
+ * @returns The address of the server.
+ */
+export const startServer = async (
+  endpoint: ParsedEndpoint,
+  config: Partial<Configuration>,
+  args: Options,
+  previous?: Port,
+): Promise<ServerAddress> => {
+  // Define the request handler for the server.
+  const serverHandler = async (
+    request: IncomingMessage,
+    response: ServerResponse,
+  ): Promise<void> => {
+    if (args['--cors']) response.setHeader('Access-Control-Allow-Origin', '*');
+    // @ts-expect-error The `compression` library uses the Express Request type
+    // instead of the native HTTP IncomingMessage type.
+    if (!args['--no-compression']) await compress(request, response);
```

Comment:
This can be better solved with:
```suggestion
    type ExpressRequest = Parameters<typeof compress>[0];
    type ExpressResponse = Parameters<typeof compress>[1];
    if (!args['--no-compression']) await compress(request as ExpressRequest, response as ExpressResponse);
```

As disabling TS for the line will hide other errors.